### PR TITLE
Convert output layer of a (general) parameterized model to a spectral library

### LINKED ISSUE #97: Use parquet instead of tsv
Due to the size of the data involved, we should update tooling to be able to read and write parquet files too. This shouldn't be too difficult in places we have used pandas (or dask) since we can also dump to parquet with little trouble.

### PR DESCRIPTION
**Pull request by raar1**
Please see https://github.com/aiproteomics/aiproteomics/pull/96#issuecomment-2485639223 further down for a more up-to-date description of this PR

Old PR description:
Adds:
* ModelParam class to hold basic parameters for the given msms model (i.e. sequence length, list of allowed ions, list of neutral losses etc)
* Code to generate the corresponding fragment info for each intensity in the predicted output layer, given an appropriate ModelParam object
* Generate a spectral library entry (as pandas dataframe) for a given output layer, given the sequence, model params etc. This can then be written as e.g. `tsv` or `parquet` as desired.

Once tested/polished, this will completely replace the adapted `prosit` stuff currently in the `e2e/` directory. It allows arbitrary length sequences, ion types, num charges, neutral losses, and phosphorylation of S, T and Y.

Fixes #97

### CODE REVIEW

**Comment by raar1** [comment]
This now needs some refactoring and resolution of #95. Then it can be reviewed and merged.

**Comment by raar1** [comment]
@dsmits 

OK, this PR is now much more substantial than at first.

This also now introduces an `AIProteomicsModel` class that holds the 3 things required for models we look at in this project:
* INPUT: A `SequenceMapper` instance, that is responsible for mapping peptide input strings (with UniMod modifications) to an integer `numpy` array that is an input layer to the AI model.
* OUTPUT: A `ModelParams` child instance that contains the parameters and methods necessary to map the output layer of the AI model to the desired form:
    * A `ModelParamsMSMS` object will map an output layer to a spectral library entry (returned as a dataframe).
    * A `ModelParamsRT` object will map the output layer to a (rescaled) normalized retention time
    * A `ModelParamsCCS` object will map the output to an ion mobility value (this one is still under construction)
* AI MODEL: An instance of the (most likely) neural network model to be used to transform the inputs to the outputs... This currently has to be a `keras` model object.

The AI model generators are found in `aiproteomics/core/models.py`. Currently there is just one (dummy) function that returns a keras model for testing purposes. But this is where we could maybe add other model building functions (such as the ones Djura is working on).

The purpose of the overall `AIProteomicsModel` object is to keep all these matching objects together in one place, and allow saving/reloading of the whole model.

The `AIProteomicsModel.to_dir()` method writes all the info to a directory, including a `config.json` file with metadata and the AI model itself (currently stored in `onnx` format, but all these choices can be discussed).

To see examples of it being used, there is a `test_aiproteomics_model.py` script in the `demo/` directory of this branch. Please have a play with it and let's discuss the design :)

**Comment by raar1** [comment]
OK, there is quite a big rewrite in this PR already so I'm going to merge it now. The refactoring and the new infrastructure is complete. Only the last work by Djura (training scripts for the MSMS Phospho model still should be added to the repo).

In the coming days I will also add a bit more documentation and demo scripts, and tidy a bit.
